### PR TITLE
Fix expression parity edge cases found reviewing #252

### PR DESCRIPTION
**Pull request by lox**
Follow-ups from a review of #252. Five behavioral gaps against GitHub's evaluator, each with regression tests.

## Nested matrix references failed validation (corpus regression)

#252 added `validateCompileExpressionNode` as a pre-evaluation gate in `EvaluateCompile` that restricted `matrix` references to one path segment. `runs-on: ${{ matrix.config.os }}` resolved before the merge and failed with `unsupported compile-time reference "matrix.config.os"` after it, flipping a previously admitted corpus workflow to incompatible. Matrix values may be objects or arrays (include entries and object lists), so nested references are valid; `resolveCompileReference` already walked them.

Runtime templates and conditions rejected nested matrix references on both sides of #252, which meant a workflow whose compile validation passed could still fail at run-job time. Both now walk nested paths as well, with missing or scalar intermediate segments yielding null to match GitHub.

| Surface | Before #252 | After #252 | Now |
|---|---|---|---|
| `runs-on: ${{ matrix.config.os }}` | resolved | rejected | resolved |
| `run: ${{ matrix.config.name }}` | fails at run-job | fails at run-job | evaluates |
| `if: matrix.config.name == 'linux'` | rejected | rejected | evaluates |

## Nested `inputs` receiver properties were rewritten

`SubstituteCompileInputs` matched any `inputs . <ident>` token run, so property chains on other receivers were corrupted:

| Template | Before | After |
|---|---|---|
| `${{ github.event.inputs.debug }}` | `${{ github.event.true }}` | unchanged |
| `${{ steps.inputs.name }}` | `${{ steps.'prod' }}` | unchanged |
| `${{ inputs.debug }}` | `${{ true }}` | `${{ true }}` |

A preceding `.` token now marks the sequence as a property access, not the workflow-call inputs context.

## Bracket-form `github` condition references failed at evaluation

`github['event_name'] == 'push'` passed `ValidateCondition` (validation treats `github` as a static-authority root) but errored at `EvaluateCondition` because `resolveConditionRoot` had no `github` case. It now returns the GitHub context, or the existing unavailable-context error when the context is nil.

## `format()` accepted placeholders GitHub rejects

`strconv.Atoi` accepts signs, so `format('{+0}', x)` and `format('{-0}', x)` compiled locally but fail on GitHub, whose parser requires bare digits. Placeholder indexes are now digit-only; `{0}` and `{01}` still work.

## `fromJSON` duplicate keys diverged from GitHub

For `fromJSON('{"a":1,"A":2}')`, GitHub's `DictionaryContextData` is case-insensitive at construction: the later value replaces the earlier one and the first spelling survives, so `.a`, `.A`, and `['A']` all yield `2` ([JTokenExtensions.ToPipelineContextData](https://github.com/actions/runner/blob/main/src/Sdk/DTPipelines/Pipelines/ContextData/JTokenExtensions.cs#L34-L43), [DictionaryContextData setter](https://github.com/actions/runner/blob/main/src/Sdk/DTPipelines/Pipelines/ContextData/DictionaryContextData.cs#L105-L126)). We errored with "ambiguous properties" instead. `decodeJSONValue` now decodes token by token (a plain map decode loses the key order the semantics depend on) and collapses case-insensitive duplicates the same way. The ambiguity error in `objectValue` remains as a safety net for caller-built context maps and drops its misleading "compile-time" prefix, since the helper also serves condition evaluation.

## Not changed

- Bracket-form `runner['os']` stays rejected at validation; only `github`/`secrets` are static-authority roots there, and GitHub-hosted parity for it was not part of the review findings.
- The timeout-minutes coercion switch in `internal/runtime/job.go` stays separate from `githubNumber`, which coerces strings and booleans that timeout handling must reject.

## Validation

`go build`, `go test ./...`, `go test -race ./internal/expression/...`, `go vet`, `golangci-lint run` (0 issues), and `mise run smoke:local` all pass.

### CODE REVIEW

File: internal/expression/expression.go [hunk 1]
```diff
@@ -367,16 +367,73 @@ func objectValue(value any, name string) (any, bool, error) {
...
+			}
+			key, ok := keyToken.(string)
+			if !ok {
+				return nil, fmt.Errorf("fromJSON argument is invalid JSON")
+			}
+			value, err := decodeJSONToken(decoder)
+			if err != nil {
+				return nil, err
+			}
+			for existing := range object {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/expression/expression.go]: Fixed in 788fbc6: the decoder now tracks first key spellings in a lowercase-keyed index instead of rescanning decoded keys, so insertion is near-linear. Re-measured your scenario: 50,000 distinct keys decode in ~73ms on this branch (was ~11.4s).

File: internal/expression/condition.go [hunk 2]
```diff
@@ -408,6 +415,11 @@ func evaluateConditionNode(node actionlint.ExprNode, context ConditionContext) (
 
 func resolveConditionRoot(root string, context ConditionContext) (any, error) {
 	switch strings.ToLower(root) {
+	case "github":
+		if context.GitHub == nil {
+			return nil, fmt.Errorf("condition context %q is unavailable", root)
+		}
+		return context.GitHub, nil
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/expression/condition.go]: Fixed in 788fbc6: static bracket references to authority contexts (`github`, `secrets`) now route through the normal reference resolver in the evaluator, mirroring the validator's static-authority rule, and the `github` case is removed from `resolveConditionRoot`. Whole and dynamic access (`if: github`, `if: github[inputs.key]`) fails closed at evaluation even on the unvalidated composite-action path; regression tests cover both shapes.

File: internal/expression/expression.go [hunk 3]
```diff
@@ -367,16 +367,74 @@ func objectValue(value any, name string) (any, bool, error) {
...
+			}
+			key, ok := keyToken.(string)
+			if !ok {
+				return nil, fmt.Errorf("fromJSON argument is invalid JSON")
+			}
+			value, err := decodeJSONToken(decoder)
+			if err != nil {
+				return nil, err
+			}
+			lower := strings.ToLower(key)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/expression/expression.go]: Fixed in e824016: index keys are now canonicalized under the same fold equivalence as lookup — every rune maps to the smallest rune in its `unicode.SimpleFold` orbit, so keys collapse exactly when `EqualFold` considers them equal. Your `Σ`/`ς` case is a regression test (`fromJSON('{"Σ":1,"ς":2}')['Σ']` yields 2), and 50,000 distinct keys still decode in ~90ms.

**Comment by sj26** [comment]
Released in [v0.14.0](https://github.com/buildkite/buildkite-gha/releases/tag/v0.14.0)
